### ISSUE TITLE
Clean up after a driver exits.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#376]
Currently, if a driver submits a ton of long running tasks and then exits, those tasks will continue to be scheduled and will continue to execute. To see this, do the following.

1. Start a Ray cluster on one node.

    ```
    ./scripts/start_ray.sh --head --node-ip-address=127.0.0.1 --redis-port=6379
    ```

2. Connect a driver and submit a bunch of expensive tasks.

```python
import ray
import numpy as np

ray.init(redis_address="127.0.0.1:6379")

@ray.remote
def f():
  x = np.zeros((1000, 1000))
  for _ in range(10):
    np.dot(x, x)

[f.remote() for _ in range(1000)]
```

3. Then after the tasks have been submitted, kill the driver (e.g., with control-D to exit from Python).

4. Open up `top` or `htop` and see that the workers are busy running tasks.

When a driver exits, we should do the following:

- Don't execute any new tasks belonging to that driver.
- Kill any workers that are still executing tasks for that driver.
- Kill any actors that belong to that driver. This also requires releasing the resources associated with that actor.

We could do additional cleanup of state in Redis and in the workers (e.g., remote function definitions).

**Comment by robertnishihara** [comment]
Thinking about the implentation, the local scheduler can determine that the driver has exited when the driver closes its connection to the local scheduler.

**Comment by richardliaw** [comment]
For actors, running the following `script.py`: 
```
import ray
import time
import sys
from datetime import datetime
import os

@ray.actor
class Training():
    def loop_it(self):
        while True:
            time.sleep(0.5)
            with open("./new.txt", "w") as f:
                f.write("woke %s..." % datetime.now())

def main():
    t = Training()
    t.loop_it()

if __name__ == '__main__':
    ray.init(redis_address=sys.argv[1])
    main()
    for i in range(10):
        time.sleep(0.5)
```
will cause `./new.txt` to be continuously updated even after the script finishes.

**Comment by robertnishihara** [comment]
This should be addressed by #462.

**Comment by raopku** [comment]
@robertnishihara 
After the pull request, I have install the newest version of ray.
When a driver disconnects from ray cluster, the Actors and Works will be killed as we expected.
However ,when i start a new driver which connects to the same ray cluster (have the same redis address as before), new actors are always waiting to run.

```
import ray
ray.init(redis_address="9.9.9.9:6379")
agent_count = 8

actor = ray.actor(num_gpus=1)(MyActor)
agents =[actor() for _ in range(agent_count)]
for i in range(agent_count):
    agents[i].run()
```

kill the driver, connect to the same ray cluster

```
import ray
ray.init(redis_address="9.9.9.9:6379")
agent_count = 8

actor = ray.actor(num_gpus=1)(MyActor)
agents =[actor() for _ in range(agent_count)]
```
agents are not running
```
for i in range(agent_count):
    agents[i].run()
```

**Comment by robertnishihara** [comment]
Thanks for trying it out! I don't fully understand the question. Could you provide a full script that I can run which does not behave as expected? And say what output you expect to see?

**Comment by raopku** [comment]
I use the jupyter notebook. Jypyter notebook can restart the kernel, which can be viewed as kill the old driver and restart a new driver.
```
import os
os.getpid()
_output:21238_ 
restart kernel
import os
os.getpid()
_output:23021_
```
here is my code, you can use it to reproduce the error.
```
from __future__ import division
import warnings
import ray
from keras.models import model_from_config

class MyActor(object):

    def __init__(self, model_config):

 
        import os

        os.environ['CUDA_VISIBLE_DEVICES'] = ','.join([str(i) for i in ray.get_gpu_ids()])
        try:
            model = model_from_config(model_config, custom_objects={})
        except:
            print('No enough GPUs to start Actor')
            self.exit()
        self.model = model
    
    def fit(self):
        import numpy as np
        for i in range(10000):
            
            self.model.predict_on_batch(np.zeros((10,1,4)))       
ray.init(redis_address="9.91.9.123:6379")
import numpy as np
import gym
import ray
from keras.models import Sequential

from keras.layers import Dense, Activation, Flatten
from keras.optimizers import Adam


model = Sequential()
model.add(Flatten(input_shape=(1,) + (4,)))
model.add(Dense(16))
model.add(Activation('relu'))
model.add(Dense(16))
model.add(Activation('relu'))
model.add(Dense(16))
model.add(Activation('relu'))
model.add(Dense(2))
model.add(Activation('linear'))


config = {
    'class_name': model.__class__.__name__,
    'config': model.get_config(),
}

agent_count = 8

gpu_actor = ray.actor(num_gpus=1)(MyActor)
agents =[gpu_actor(config) for _ in range(agent_count)]
for i in range(agent_count):
    agents[i].fit()

```

Fisrt Step: start ray on two nodes
Before i execute my code. No one takes up GPU.
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 0000:06:00.0     Off |                    0 |
| N/A   48C    P8    27W / 149W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 0000:07:00.0     Off |                    0 |
| N/A   34C    P8    30W / 149W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 0000:85:00.0     Off |                    0 |
| N/A   55C    P8    27W / 149W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 0000:86:00.0     Off |                    0 |
| N/A   40C    P8    32W / 149W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

```

Execture my code, you can see some process is using GPU.

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 0000:06:00.0     Off |                    0 |
| N/A   53C    P0    54W / 149W |  11283MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 0000:07:00.0     Off |                    0 |
| N/A   42C    P0    70W / 149W |  11283MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 0000:85:00.0     Off |                    0 |
| N/A   60C    P0    60W / 149W |  11283MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 0000:86:00.0     Off |                    0 |
| N/A   47C    P0    74W / 149W |  11283MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      3508    C   .../anaconda2/envs/tensorflow_rao/bin/python 10897MiB |
|    0      5865    C   python                                         382MiB |
|    1      3508    C   .../anaconda2/envs/tensorflow_rao/bin/python 10867MiB |
|    1      5866    C   python                                         412MiB |
|    2      3508    C   .../anaconda2/envs/tensorflow_rao/bin/python 10869MiB |

```


Restart the kernel, which means kill the old driver and restart a new driver.
Some process is still taking up GPU.(pid = 5865)

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 0000:06:00.0     Off |                    0 |
| N/A   57C    P0    55W / 149W |    386MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 0000:07:00.0     Off |                    0 |
| N/A   41C    P8    30W / 149W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 0000:85:00.0     Off |                    0 |
| N/A   64C    P0    61W / 149W |    414MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 0000:86:00.0     Off |                    0 |
| N/A   50C    P0    74W / 149W |    416MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      5865    C   python                                         382MiB |
|    2      5867    C   python                                         410MiB |
|    3      5868    C   python                                         410MiB |
+-----------------------------------------------------------------------------+

```

you can see process 5865 is an worker of ray. 
```
dl@dl123:~$ ps -ef | grep 5865
dl        5865  2001  8 07:49 pts/12   00:00:17 python /home/dl/.local/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/workers/default_worker.py --node-ip-address=9.91.9.123 --object-store-name=/tmp/plasma_store62295397 --object-store-manager-name=/tmp/plasma_manager5130114 --local-scheduler-name=/tmp/scheduler53382161 --redis-address=9.91.9.123:6379 --actor-id ca3fb5709f41b2b9ae84835edc406d73a935ce0a
```


After start a new driver , i execute my code again. You can see more and more process is taking up GPU.
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 0000:06:00.0     Off |                    0 |
| N/A   59C    P0    55W / 149W |  11283MiB / 11439MiB |     14%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 0000:07:00.0     Off |                    0 |
| N/A   36C    P0    70W / 149W |  11283MiB / 11439MiB |     14%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 0000:85:00.0     Off |                    0 |
| N/A   70C    P0    62W / 149W |  11283MiB / 11439MiB |     14%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 0000:86:00.0     Off |                    0 |
| N/A   53C    P0    75W / 149W |  11283MiB / 11439MiB |     14%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      5865    C   python                                         382MiB |
|    0      9020    C   .../anaconda2/envs/tensorflow_rao/bin/python 10531MiB |
|    0     11866    C   python                                         363MiB |
|    1      9020    C   .../anaconda2/envs/tensorflow_rao/bin/python 10867MiB |
|    1     11867    C   python                                         412MiB |
```

repeated restart kernel and execute code several times
Finally, i got the error log:

```
Remote function fit failed with:

Traceback (most recent call last):
  File "/home/dl/.local/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/worker.py", line 1848, in process_task
    worker.actors[task.actor_id().id()], *arguments)
  File "<ipython-input-1-99bab00c23f0>", line 28, in fit
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/models.py", line 710, in predict_on_batch
    return self.model.predict_on_batch(x)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/engine/training.py", line 1279, in predict_on_batch
    outputs = self.predict_function(ins)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 1096, in __call__
    updated = session.run(self.outputs + [self.updates_op], feed_dict=feed_dict)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 766, in run
    run_metadata_ptr)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 964, in _run
    feed_dict_string, options, run_metadata)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1014, in _do_run
    target_list, options, run_metadata)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1034, in _do_call
    raise type(e)(node_def, op, message)
InvalidArgumentError: Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[Node: Reshape = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/gpu:0"](_recv_flatten_input_1_0/_1, pack)]]
	 [[Node: add_3/_7 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_20_add_3", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

Caused by op u'Reshape', defined at:
  File "/home/dl/.local/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/workers/default_worker.py", line 61, in <module>
    ray.worker.main_loop()
  File "/home/dl/.local/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/worker.py", line 1930, in main_loop
    process_task(task)
  File "/home/dl/.local/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/worker.py", line 1848, in process_task
    worker.actors[task.actor_id().id()], *arguments)
  File "<ipython-input-1-99bab00c23f0>", line 18, in __init__
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/models.py", line 189, in model_from_config
    return layer_from_config(config, custom_objects=custom_objects)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/utils/layer_utils.py", line 34, in layer_from_config
    return layer_class.from_config(config['config'])
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/models.py", line 1056, in from_config
    model.add(layer)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/models.py", line 292, in add
    layer.create_input_layer(batch_input_shape, input_dtype)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/engine/topology.py", line 374, in create_input_layer
    self(x)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/engine/topology.py", line 517, in __call__
    self.add_inbound_node(inbound_layers, node_indices, tensor_indices)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/engine/topology.py", line 571, in add_inbound_node
    Node.create_node(self, inbound_layers, node_indices, tensor_indices)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/engine/topology.py", line 155, in create_node
    output_tensors = to_list(outbound_layer.call(input_tensors[0], mask=input_masks[0]))
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/layers/core.py", line 440, in call
    return K.batch_flatten(x)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 862, in batch_flatten
    x = tf.reshape(x, tf.pack([-1, prod(shape(x)[1:])]))
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2448, in reshape
    name=name)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 759, in apply_op
    op_def=op_def)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2240, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/home/dl/anaconda2/envs/tensorflow_rao/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1128, in __init__
    self._traceback = _extract_stack()

InvalidArgumentError (see above for traceback): Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[Node: Reshape = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/gpu:0"](_recv_flatten_input_1_0/_1, pack)]]
	 [[Node: add_3/_7 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_20_add_3", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]



You can inspect errors by running

    ray.error_info()

If this driver is hanging, start a new one with

    ray.init(redis_address="9.91.9.123:6379")
```

**Comment by robertnishihara** [comment]
This is a bug, thanks for catching this! The actors are not actually being killed. I'm reopening this issue.

It can be reproduced more easily as follows. First start Ray as follows.

```
./scripts/start_ray.sh --head --node-ip-address=127.0.0.1 --redis-port=6379
```

Then run the following script.

```python
import ray
import numpy as np

ray.init(redis_address="127.0.0.1:6379")

@ray.actor
class Foo(object):
  def __init__(self):
    x = np.zeros((1000, 1000))
    while True:
      np.dot(x, x)

actors = [Foo() for _ in range(8)]
```

Open up `htop` to se that all the cores are being used.

Exit from the driver, and notice that all of the cores are still being used.

**Comment by robertnishihara** [comment]
Ok, this should be fixed now by #490. The fix was actually quite simple (see https://github.com/ray-project/ray/pull/490/commits/fef21c6d6792ec590194bd2c2cec19c43903e7f6). There was an incorrect `break` in the code which meant we only killed at most one worker/actor.

**Comment by robertnishihara** [comment]
@raopku, let me know if you continue to see the problem!
